packages/errors/lib/operational-error.js

```diff
@@ -44,4 +86,30 @@ module.exports = class OperationalError extends Error {
 		// case as we're manually casting `undefined` to a Boolean
 		return Boolean(error.isOperational);
 	}
-};
+
+	/**
+	 * Normalize a machine-readable error code.
+	 *
+	 * @access private
+	 * @param {String} code
+	 *     The error code to normalize.
+	 * @returns {String}
+	 *     Returns the normalized error code.
+	 */
+	static normalizeErrorCode(code) {
+		return code
+			.trim()
+			.toUpperCase()
+			.replace(/[^a-z0-9_]+/gi, '_');
+	}
+}
+
+/**
+ * @typedef {Object} OperationalErrorData
+ * @property {String} [code]
+ *     A machine-readable error code which identifies the specific type of error.
+ * @property {String} [message]
+ *     A human readable message which describes the error.
+ */
```

Comment:
**thought**: in typescript you could allow arbitrary extra stuff in this object either with an index signature:

```typescript
type OperationalErrorData = {
  code: string
  message: string
  [key: string]: any
}
```

or via an intersection type:

```typescript
type OperationalErrorData = {
  code: string
  message: string
} & Record<string, any>
```

from a quick google i don't know if either of these are possible in JSDoc but might be worth another look, requiring `@ts-ignore` isn't ideal

Reply:
also **suggestion**: it feels weird to me to have this definition at the end of the file? idk why but i think it would flow better with this at the top

Reply:
Adding these types _are_ possible in JSDoc with the following:

```js
/**
 * @typedef {Object<String, String>} OperationalErrorData
 */
```

Or in the class with

```
/**
 * @param {(String|OperationalErrorData & Object<String, String>)} [data = {}]
 */
```

However that results in VS Code's autocompletion saying that the `data` parameter is just `any` because it infers that an object which can have any string key is basically the same. So I'm torn between a ts-ignore but proper type-hinting for the `message` and `code` properties, or no type hinting but technically correct typing.

Maybe correct typing is the right way to go, as we're documenting everything as well?

Reply:
this is the type that shows up for me if i try your second suggestion, which looks correct:

<img width="537" alt="Screenshot 2022-05-06 at 13 26 12" src="https://user-images.githubusercontent.com/631757/167130992-e694d430-9aa1-4eb0-84da-556b2a90266c.png">

Reply:
actually yeah this is working perfectly for me?

<img width="1680" alt="Screenshot 2022-05-06 at 14 09 29" src="https://user-images.githubusercontent.com/631757/167137747-1b37a83d-63e5-4ea3-891b-4daba5a7e27a.png">

Reply:
actually yeah this is working perfectly for me:

<img width="176" alt="Screenshot 2022-05-06 at 14 10 22" src="https://user-images.githubusercontent.com/631757/167137873-31cec760-c32b-4b7d-89a5-a4c13f866488.png">

Reply:
ah, nice! wasn't sure to what extent typescript syntax would work here. `{[key: string]: any}` is i believe equivalent to `Record<string, any>` which imo looks nicer if it works?

Reply:
Nice! This also works. Thanks for all your help :)